service-search-nls/src/main/java/fi/nls/oskari/search/util/ELFGeoLocatorCountries.java

```diff
@@ -31,53 +30,118 @@
     private long lastSync;
     private String serviceURL;
     private ELFGeoLocatorSearchChannel channel;
-    private Map<String, String> countries = new HashMap<>();
+    private static Map<String, String> countryMap = new HashMap<>();
 
-    public ELFGeoLocatorCountries(ELFGeoLocatorSearchChannel channel) {
-        this.channel = channel;
+    public ELFGeoLocatorCountries() {
         serviceURL = PropertyUtil.getOptional(ELFGeoLocatorSearchChannel.PROPERTY_SERVICE_URL);
         if (serviceURL == null) {
             log.warn("ServiceURL not configured. Add property with key", ELFGeoLocatorSearchChannel.PROPERTY_SERVICE_URL);
         }
+        loadCountryMap();
     }
 
-    public Map<String, String> getCountryMap() throws IOException {
-        // not properly configured
-        if (serviceURL == null) {
-            return countries;
+    public void loadCountryMap () {
+        try {
+            // not properly configured
+            if (serviceURL == null) {
+                return ;
+            }
+
+            // synced recently
+            if(System.currentTimeMillis() < lastSync + REFRESH_INTERVAL) {
+                return ;
+            }
+            parseCountryMap(getCountryData());
+            if(countryMap.isEmpty()) {
+                log.debug("Could not get countries");
+            }
         }
-        // synced recently
-        if(System.currentTimeMillis() < lastSync + REFRESH_INTERVAL) {
-            return countries;
+        catch (Exception e) {
+            log.debug(e, "Could not get countries");
         }
+        lastSync = System.currentTimeMillis();
+    }
 
-        // Start from scratch
+    public String getCountryData() throws IOException {
         String countriesUrl = serviceURL + DEFAULT_GETCOUNTRIES_TEMPLATE;
-        StringBuffer buf = new StringBuffer(countriesUrl);
-        log.debug("Server request: " + buf.toString());
-        HttpURLConnection conn = channel.getConnection(buf.toString());
+        log.debug("Server request: " + countriesUrl);
+        HttpURLConnection conn = IOHelper.getConnection(countriesUrl);
         String response = IOHelper.readString(conn);
-        log.debug("Server response: " + response);
+        return response;
+    }
+
+    public void parseCountryMap(String response) {
+        if (response == null || response.isEmpty()) {
```

Comment:
Should return the parsed result or be made private.

Reply:
This refactoring was made to make unit tests possible. Couldn't think of a better solution at the moment.

Reply:
This method should return a new Map containing the parsed result. Consider retuning a null or throwing an exception if something fails.

Reply:
How would we set the country map in that case, so we can handle unit tests also.

Reply:
Change the declaration to `public Map<String, String> parseCountryMap(String response);` instead in loadCountryMap() change the reference of the class variable to use the newly parsed Map. In Unit tests validate that the Map looks like what it should